pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/PulsarAuthorizationProvider.java

```diff
@@ -305,37 +305,27 @@ public void initialize(ServiceConfiguration conf, PulsarResources pulsarResource
     @Override
     public CompletableFuture<Void> grantPermissionAsync(NamespaceName namespaceName, Set<AuthAction> actions,
             String role, String authDataJson) {
-        CompletableFuture<Void> result = new CompletableFuture<>();
-
         try {
             validatePoliciesReadOnlyAccess();
         } catch (Exception e) {
-            result.completeExceptionally(e);
-            return result;
+            return FutureUtil.failedFuture(e);
         }
 
-        try {
-            pulsarResources.getNamespaceResources().setPolicies(namespaceName, policies -> {
-                policies.auth_policies.getNamespaceAuthentication().put(role, actions);
-                return policies;
-            });
-            log.info("[{}] Successfully granted access for role {}: {} - namespace {}", role, role, actions,
-                    namespaceName);
-            result.complete(null);
-        } catch (NotFoundException e) {
-            log.warn("[{}] Failed to set permissions for namespace {}: does not exist", role, namespaceName);
-            result.completeExceptionally(new IllegalArgumentException("Namespace does not exist" + namespaceName));
-        } catch (BadVersionException e) {
-            log.warn("[{}] Failed to set permissions for namespace {}: concurrent modification", role, namespaceName);
-            result.completeExceptionally(new IllegalStateException(
-                    "Concurrent modification on metadata: " + namespaceName + ", " + e.getMessage()));
-        } catch (Exception e) {
-            log.error("[{}] Failed to get permissions for namespace {}", role, namespaceName, e);
-            result.completeExceptionally(
-                    new IllegalStateException("Failed to get permissions for namespace " + namespaceName));
-        }
+        CompletableFuture<Void> future = pulsarResources.getNamespaceResources()
+                .setPoliciesAsync(namespaceName, policies -> {
+                    policies.auth_policies.getNamespaceAuthentication().put(role, actions);
+                    return policies;
+                }).thenRun(() -> {
+                    log.info("[{}] Successfully granted access for role {}: {} - namespace {}", role, role, actions,
+                            namespaceName);
+                });
 
-        return result;
+        future.exceptionally(ex -> {
+            log.error("[{}] Failed to set permissions for role {} namespace {}", role, role, namespaceName, ex);
```

Comment:
@mattisonchao - great catch, we should use `getCause` here. I'll revisit the log line too. Both of these are copied closely from https://github.com/apache/pulsar/pull/12515, so I'll take a look at modifying that code too. Thanks!

Reply:
Regarding the log line, I think the lack of clarity is because the API is a bit inconsistent by using policies and permission interchangeably. There are several instances in the file that indicate `Failed to set permissions` or `Failed to get permissions`. I think we should leave it as is for now and possibly change it in another PR. Let me know what you think. Thanks.

Reply:
Agree with you +1 !